# Dimension Reduction  차원 축소 

__< What is & Why we use dimension reduction ? >__<br><br>
매우 많은 피처로 구성된 다차원 데이터 세트의 차원을 축소해 새로운 차원의 데이터 세트를 생성하는 것<br><br>
피처가 많은 데이터 세트의 경우, 상대적으로 적은 차원에서 학습된 모델보다 예측 신뢰도가 떨어진다.<br>
그리고 개별 피처 간에 상관 관계가 높을 가능성이 커지기 때문에 '다중 공선성 문제' 로 이어져 예측 성능이 저하된다. <br><br>
이러한 문제로 다차원의 피처를 차원 축소하면 직관적으로 데이터를 해석 할 수 있을 뿐만 아니라 시각적으로 데이터의 특성을 파악하거나 학습에 필요한 처리 능력을 줄이는 등의 이점을 가져 갈 수 있다.<br><br>

__< Dimension reduction type >__
* Feature selection 
    * 특정 피처에 종속성이 강한 불필요한 피처는 아예 제거하고, 주요 피처만 선택하는 방법 


* Feature extraction
    * 기존 피처를 저차원의 중요 피처로 압축해서 추출하는 방법 
    * 기존 피처를 단순 압축이 아닌, 데이터 들을 더 잘 설명 할 수 있는 함축적인 특정을 살려 또 다른 공간으로 매핑해 추출하는 것으로,<br> 기존 피처와는 완전히 다른 값 
    * EX) 모의고사 성적 feature , 수능 성적 feature , 대학교 1 ~ 4 학년 전공 성적 feature 등을 ' 학업 성취도 ' feature 로 압축하는 것 
<br><br>   

__< Dimension reduction algorithm>__
* PCA - Principal Component Analysis
* LDA - Linear Discriminant Analysis
* SVD - Single Value Decomposition
* NMF - Non negative Matrix Factorization
<br><br>

__< Dimension reduction algorithm uses>__
1. 이미지 데이터 
* 매우 많은 픽셀로 이뤄진 이미지 데이터에서 잠재된 특성을 피처로 도출해 함축적 형태의 이미지 변환과 압축을 수행 할 수 있다.<br>
  이 때, 변환된 이미지는 분류 수행 시에 과적합 영향력이 작아져서 오히려 원본 이미지로 예측하는 것보다 예측 성능을 더 끌어 올릴 수 있다.

2. 텍스트 분석
* 문서 내 단어들의 구성에서 숨겨져 있는 Semantic 의미나 Topic 을 잠재 요소로 간주하고 이를 찾아 낼 수 있다.<br>
  SVD , NMF 가 여기에 사용된다.


## PCA 

__< What is PCA ? >__<br><br>
여러 변수 = 피처 간에 존재하는 상관관계를 이용해 이를 대표하는 주성분 ( principal component ) 을 추출해 차원을 축소하는 기법.<br>
정보 유실이 최소화 되야 하기 때문에 PCA 는 가장 높은 분산을 가지는 데이터의 축 = 주성분 을 찾아 이 축으로 차원 축소를 한다. <br>
단, 분산이 데이터 특성을 가장 잘 나타낸 다는 가정 하 <br><br>
이 [그림](https://liorpachter.files.wordpress.com/2014/05/pca_figure1.jpg?w=490&h=490) 을 통해 간단하게 원리를 확인 해보자. *p.379 참고* 

> 평 수 ( X ) 와 주택 가격 ( Y ) 두 피처를 가지고 있는 데이터 세트 ( = 흰 원 ) 이 있다.<br><br>
    2개의 피처를 한 개의 주성분을 가진 데이터 세트로 차원 축소를 하기 위해 __데이터 변동성이 가장 큰 = 분산이 큰__ 데이터의 축 ( = 초록 선 ) 을 생성하고, 해당 축으로 데이터를 투영하는 ( = 파란 원 ) 방식이다.



이번에는 다른 [그림](https://miro.medium.com/max/602/0*PnqMbZEdnuL9yHuo.png) 을 통해서 더 자세하게 이해해 보자.

> PCA 는 제일 먼저 가장 큰 variance 기반으로 첫 번째 벡터 축을 생성하고, 다음 축은 이 벡터 축에 직각이 되는 직교 벡터로 축으로 생성한다. <br>
이렇게 생성된 벡터 축에 원본 데이터를 투영하면, __벡터 축의 개수만큼의 차원__ 으로 원본 데이터가 차원 축소 된다.

<br><br>

__< PCA from a linear algebra perspective >__<br><br>
입력 데이터의 공분산 행렬이 고유벡터와 고유값으로 분해 될 수 있으며, 이렇게 분해된 고유 벡터를 이용해 입력 데이터를 선형 변환하는 방식<br>
### <span style="color:red">중요</span> 
[__< Loner 의 학습노트 >__](https://wiserloner.tistory.com/1085) , [__< 공돌이의 수학정리 노트 >__](https://angeloyeo.github.io/2019/07/17/eigen_vector.html), [__< 다크 프로그래머 >__](https://darkpgmr.tistory.com/105) 를 통해, <br>자세한 개념과 수학적 증명을 담은 해당 링크를 참고하자!<br><br>

------------

### 간략한 정리
 * __선형 변환__<br>
vector 에 matrix 를 내적해서 크기 or 방향이 변하는 선형 형태의 vector 를 만들어 주는 것<br><br>
 * __고유값 , 고유 벡터__  <br>
선형 변환을 할 때, 크기만 변하고 방향이 변하지 않을 수 있다. 이 때, <br>
> Ax=λx ( x는 vector )

    입력 벡터 x 를 행렬 A 로 선형 변환 시킨 결과가 입력 벡터 x 의 상수 배<br>
    즉, 이 때 vector x  = ' 고유 벡터' , 숫자 λ = ' 고유값 ' <br><br>

 * __공분산 = Covariance__ <br>
> COV(x,y) = E{(x-E(x))* (y-E(y))} <br><br> 전체 관측값의 평균 { ( 특정 관측값 - X값 평균 ) * ( 특정 관측값 - Y값 평균 ) } 

    두 피처 간의 상관 관계를 나타내는 값 <br><br>
 * __공분산 행렬__ <br>
공분산을 나타내는 행렬 = 특정 피처들의 분산이 얼마나 닮았는가 를 나타내는 행렬 <br><br>
 * __공분산 행렬에서 고유 벡터와 고유값의 의미__<br>
고유 벡터 = 데이터가 어떤 방향으로 분산 되어 있는지 <br>
고유 값 = 고유 벡터의 크기 = 분산의 크기 <br><br>

<span style="color:red">즉, PCA 는 분산이 큰 = 고유 값이 큰 순서대로 고유 벡터를 정렬하면서 결과적으로 중요한 순서대로 주성분을 구하는 것. </span>

-----------------


__< PCA , order of execution >__
1. 입력 데이터 세트의 공분산 행렬을 생성
2. 공분산 행렬의 고유벡터와 고유값을 계산
3. 고유값이 가장 큰 순으로 K개 ( PCA 변환 차수 ) 만큼 고유벡터를 추출
4. 고유값이 가장 큰 순으로 추출된 고유벡터를 이용해 새롭게 입력 데이터를 변환 
  
<br><br>
이제 내장 dataset 붓꽃 데이터를 이용해서 PCA 를 구현해 보자.

In [49]:
from sklearn.datasets import load_iris
import pandas as pd 
import plotly.graph_objects as go

iris = load_iris()

iris_df = pd.DataFrame(data=iris.data,columns=iris.feature_names)
iris_df['target']=iris.target 
iris_df.columns=['sepal_length','sepal_width','petal_length','petal_width','target']

iris_df

sepal_length  sepal_width  petal_length  petal_width  target
0             5.1          3.5           1.4          0.2       0
1             4.9          3.0           1.4          0.2       0
2             4.7          3.2           1.3          0.2       0
3             4.6          3.1           1.5          0.2       0
4             5.0          3.6           1.4          0.2       0
..            ...          ...           ...          ...     ...
145           6.7          3.0           5.2          2.3       2
146           6.3          2.5           5.0          1.9       2
147           6.5          3.0           5.2          2.0       2
148           6.2          3.4           5.4          2.3       2
149           5.9          3.0           5.1          1.8       2

[150 rows x 5 columns]

먼저, 두 개의 피처 sepal length 와 sepal width 를 이용해서 2차원 분포로 시각화 해본다.<br><br>
참고 : [Plotly text & annotation 으로 그래프 구체화 하기](https://plotly.com/python/text-and-annotations/)

In [50]:
graph = go.Figure()

for i in range(0,3):
    
    graph.add_trace(
    go.Scatter(x=iris_df[iris_df['target']==i]['sepal_length'],y=iris_df[iris_df['target']==i]['sepal_width'],mode='markers' ,
               name = iris.target_names[i])
    )

graph.update_xaxes(title_text="sepal_length")    
graph.update_yaxes(title_text="sepal_width")

graph.show()

이제는 PCA 로 4개 속성을 2개로 압축한 뒤 앞의 예제와 비슷하게 2개의 PCA 속성으로 붓꽃 데이터의 품종 분포를 2차원으로 시각화 해보자.<br>
PCA 를 적용하기 전에 각 피처를 __동일한 스케일로 변환__ 시켜주는 과정이 필요하다.

In [51]:
from sklearn.preprocessing import StandardScaler

# target 값은 제외하고 스케일링 
iris_scaled = StandardScaler().fit_transform(iris_df.iloc[:,:-1])

이제 PCA 를 적용해서 차원을 축소 시켜준 뒤에 새로운 2개의 속성으로 변환된 데이트 세트를 시각화 한다.

In [52]:
from sklearn.decomposition import PCA

# n_components = PCA 로 변환할 차원의 수 
pca = PCA(n_components=2)

pca.fit(iris_scaled)
iris_pca = pca.transform(iris_scaled)

iris_pca.shape

(150, 2)

In [53]:
pca_columns = ['pca_component1','pca_component2']

iris_df_pca = pd.DataFrame(iris_pca,columns=pca_columns)
iris_df_pca['target'] = iris.target

iris_df_pca

pca_component1  pca_component2  target
0         -2.264703        0.480027       0
1         -2.080961       -0.674134       0
2         -2.364229       -0.341908       0
3         -2.299384       -0.597395       0
4         -2.389842        0.646835       0
..              ...             ...     ...
145        1.870503        0.386966       2
146        1.564580       -0.896687       2
147        1.521170        0.269069       2
148        1.372788        1.011254       2
149        0.960656       -0.024332       2

[150 rows x 3 columns]

In [54]:
pca_graph = go.Figure()

for i in range(0,3):
    
    pca_graph.add_trace(
    go.Scatter(x=iris_df_pca[iris_df_pca['target']==i]['pca_component1'],
               y=iris_df_pca[iris_df_pca['target']==i]['pca_component2'],mode='markers' ,
               name = iris.target_names[i])
    )

pca_graph.update_xaxes(title_text="pca_component1")    
pca_graph.update_yaxes(title_text="pca_component2")

pca_graph.show()

PCA 변환 후에도 pca_component1 축을 기반으로 Setosa 품종은 명확하게 구분이 가능하다. Versicolor 와 Virginica 역시 pca_component1 축을 기준으로 서로 겹치는 부분이 일부 존재하지만, PCA 변환 전보다는 잘 구분되었다. <br><br>
그 이유는 pca_component1 이 원본 데이터의 변동성을 잘 반영했기 때문이다. <br>
이처럼 새롭게 만들어진 두 피처 = 속성 이 각 각 원본 데이터의 변동성을 얼마나 잘 반영했는지 확인 하기 위해서 PCA 객체는 아래와 같은 속성을 지원한다.

In [7]:
# pca_component1 , pca_component2
print(pca.explained_variance_ratio_)

[0.72962445 0.22850762]


따라서 PCA 를 2개 요소로만 변환해도 원본 데이터의 변동성을 95% 나 설명 할 수 있다. <br><br>
이번에는 원본 붓꽃 dataset = iris_df 와 PCA 변환된 dataset = iris_df_pca 를<br>
RandomForestClassifier 알고리즘과 cross_val_score 교차 검증을 통해 평가 지표 scoring 을 이용해서 예측 성능을 계산해보자.

In [8]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score
import numpy as np

rf_clf =  RandomForestClassifier(random_state=166)

scores = cross_val_score(rf_clf,iris.data,iris.target,scoring='accuracy',cv=3)

print('원본 붓꽃 Dataset 교차 검증 개별 정확도: ',scores)
print('원본 붓꽃 Dataset 평균 정확도: ',np.mean(scores))

원본 붓꽃 Dataset 교차 검증 개별 정확도:  [0.98 0.94 0.96]
원본 붓꽃 Dataset 평균 정확도:  0.96


In [9]:
scores = cross_val_score(rf_clf,iris_df_pca.iloc[:,:-1].values,iris_df_pca['target'].values,scoring='accuracy',cv=3)

print('PCA 변환된 붓꽃 Dataset 교차 검증 개별 정확도: ',scores)
print('PCA 변환된 붓꽃 Dataset 평균 정확도: ',np.round(np.mean(scores),2))

PCA 변환된 붓꽃 Dataset 교차 검증 개별 정확도:  [0.88 0.88 0.9 ]
PCA 변환된 붓꽃 Dataset 평균 정확도:  0.89


원본 데이터 세트 대비 예측 정확도는 PCA 변환 차원 개수에 따라 예측 성능이 떨어질 수 밖에 없다. <br><br>
두 결과를 보면 약 0.7 가량 정확도가 떨어짐을 알 수 있는데 물론 큰 수치긴 하나 피처의 갯수가 50% 가 된 점을 고려하면 <br> PCA 변환을 하여도 원본 데이터을 상당 부분이 유지 되었음을 알 수 있다. <br><br>

이번에는 피처의 갯수가 24개인 ' UCI Machine Learning Repository ' 에 있는 신용카드 고객 데이터 세트를 사용해서 PCA 를 진행해 보자. <br>
[Data Origin](https://archive.ics.uci.edu/ml/datasets/default+of+credit+card+clients)

In [10]:
df = pd.read_excel('credit_card_clients.xls',header=1,sheet_name='Data').iloc[0:,1:]
df

LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  PAY_4  \
0          20000    2          2         1   24      2      2     -1     -1   
1         120000    2          2         2   26     -1      2      0      0   
2          90000    2          2         2   34      0      0      0      0   
3          50000    2          2         1   37      0      0      0      0   
4          50000    1          2         1   57     -1      0     -1      0   
...          ...  ...        ...       ...  ...    ...    ...    ...    ...   
29995     220000    1          3         1   39      0      0      0      0   
29996     150000    1          3         2   43     -1     -1     -1     -1   
29997      30000    1          2         2   37      4      3      2     -1   
29998      80000    1          3         1   41      1     -1      0      0   
29999      50000    1          2         1   46      0      0      0      0   

       PAY_5  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
0         -2  ...          0          0          0         0       689   
1          0  ...       3272       3455       3261         0      1000   
2          0  ...      14331      14948      15549      1518      1500   
3          0  ...      28314      28959      29547      2000      2019   
4          0  ...      20940      19146      19131      2000     36681   
...      ...  ...        ...        ...        ...       ...       ...   
29995      0  ...      88004      31237      15980      8500     20000   
29996      0  ...       8979       5190          0      1837      3526   
29997      0  ...      20878      20582      19357         0         0   
29998      0  ...      52774      11855      48944     85900      3409   
29999      0  ...      36535      32428      15313      2078      1800   

       PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
0             0         0         0         0                           1  
1          1000      1000         0      2000                           1  
2          1000      1000      1000      5000                           0  
3          1200      1100      1069      1000                           0  
4         10000      9000       689       679                           0  
...         ...       ...       ...       ...                         ...  
29995      5003      3047      5000      1000                           0  
29996      8998       129         0         0                           0  
29997     22000      4200      2000      3100                           1  
29998      1178      1926     52964      1804                           1  
29999      1430      1000      1000      1000                           1  

[30000 rows x 24 columns]

해당 dataset 를 보면, PAY_0 피처 다음에 PAY_2 피처가 오고 있으므로 피처명을 변경한다.<br><br>
그리고 Label 로 사용할 default payment next month 의 경우 ' 다음 달 연체 여부' 를 의미하며 연체 일 경우 1, 정상 납부일 경우 0 으로 표현 되었다.<br>
피처명이 너무 기므로 임의로 짧게 줄였다. 

In [11]:
df.rename(columns={'PAY_0':'PAY_1','default payment next month':'overdue'},inplace=True)
y_target = df['overdue']
X_feature = df.drop('overdue',axis=1)

df

LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_1  PAY_2  PAY_3  PAY_4  \
0          20000    2          2         1   24      2      2     -1     -1   
1         120000    2          2         2   26     -1      2      0      0   
2          90000    2          2         2   34      0      0      0      0   
3          50000    2          2         1   37      0      0      0      0   
4          50000    1          2         1   57     -1      0     -1      0   
...          ...  ...        ...       ...  ...    ...    ...    ...    ...   
29995     220000    1          3         1   39      0      0      0      0   
29996     150000    1          3         2   43     -1     -1     -1     -1   
29997      30000    1          2         2   37      4      3      2     -1   
29998      80000    1          3         1   41      1     -1      0      0   
29999      50000    1          2         1   46      0      0      0      0   

       PAY_5  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
0         -2  ...          0          0          0         0       689   
1          0  ...       3272       3455       3261         0      1000   
2          0  ...      14331      14948      15549      1518      1500   
3          0  ...      28314      28959      29547      2000      2019   
4          0  ...      20940      19146      19131      2000     36681   
...      ...  ...        ...        ...        ...       ...       ...   
29995      0  ...      88004      31237      15980      8500     20000   
29996      0  ...       8979       5190          0      1837      3526   
29997      0  ...      20878      20582      19357         0         0   
29998      0  ...      52774      11855      48944     85900      3409   
29999      0  ...      36535      32428      15313      2078      1800   

       PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  overdue  
0             0         0         0         0        1  
1          1000      1000         0      2000        1  
2          1000      1000      1000      5000        0  
3          1200      1100      1069      1000        0  
4         10000      9000       689       679        0  
...         ...       ...       ...       ...      ...  
29995      5003      3047      5000      1000        0  
29996      8998       129         0         0        0  
29997     22000      4200      2000      3100        1  
29998      1178      1926     52964      1804        1  
29999      1430      1000      1000      1000        1  

[30000 rows x 24 columns]

이제 24개의 피처별 상관 관계를 살펴보자. 이유는 PCA 는 상관관계를 이용해 주성분을 추출한다고 언급했었다.<br><br>
[plotly expree basic](https://plotly.com/python/plotly-express/)<br>
[plotly express 색 테마 변경](https://plotly.com/python/builtin-colorscales/) <br>
[plotly express imshow](https://plotly.com/python/imshow/)

In [39]:
import plotly.express as px

graph = px.imshow(df.corr(),color_continuous_scale="blues")

graph.show()

먼저 BILL_AMT 1 ~ 6 에서 상관도가 대부분 0.9 이상 수준으로 매우 높은 것을 알 수 있으며 PAY_AMT 1 ~ 6 도 0.5 이상의 높은 상관도를 가지고 있다.<br><br>
이번에는 BILL_AMT 의 6개 속성을 2개의 component 로 PCA 변환 한 뒤 반영 변동성을 살펴보자.

In [41]:
# BILL_AMT 1 ~ 6 까지 추출 
cols_bill = ['BILL_AMT'+str(i) for i in range(1,7)]

scaler = StandardScaler()

df_cols_bill_scaled = scaler.fit_transform(X_feature[cols_bill])
pca = PCA(n_components=2)

pca.fit(df_cols_bill_scaled)

print("PCA Component 별 변동성: ",pca.explained_variance_ratio_)

PCA Component 별 변동성:  [0.90555253 0.0509867 ]


단 2개의 PCA component 만으로 6개의 피처의 변동성을 약 95% 나 설명 할 수 있다.<br><br>
이번에는 원본 데이터 세트를 RandomForest 모델 / cross_val_score 로 교차검증 /  accuracy 평가 지표  &nbsp;&nbsp;&nbsp; 를  사용 할 것이고,

* PCA 변환없이 단순 모델 학습 및 예측 성능 평가 
* 6개의 component 로 PCA 변환을 한 뒤에 모델 학습 및 예측 성능 평가 

이렇게 두 케이스로 나누어서 각자 학습을 한 뒤에 비교해보자.

In [44]:
rf_clf = RandomForestClassifier(n_estimators=300,random_state=166)

scores = cross_val_score(rf_clf,X_feature,y_target,scoring='accuracy',cv=3)

print('원본 신용카드 고객 dataset 교차 검증 개별 정확도: ',scores)
print('원본 신용카드 고객 dataset 평균 정확도: ',np.round(np.mean(scores),3))

원본 신용카드 고객 dataset 교차 검증 개별 정확도:  [0.808  0.8208 0.8229]
원본 신용카드 고객 dataset 평균 정확도:  0.817


In [48]:
scaler = StandardScaler()

X_feature_scaled = scaler.fit_transform(X_feature)
pca = PCA(n_components=6)

X_feature_scaled_pca = pca.fit_transform(X_feature_scaled)

scores = cross_val_score(rf_clf,X_feature_scaled_pca,y_target,scoring='accuracy',cv=3)

print('PCA 변환 신용카드 고객 dataset 교차 검증 개별 정확도: ',scores)
print('PCA 변환 신용카드 고객 dataset 평균 정확도: ',np.round(np.mean(scores),3))
print("PCA Component 별 변동성: ",pca.explained_variance_ratio_)

PCA 변환 신용카드 고객 dataset 교차 검증 개별 정확도:  [0.7895 0.7964 0.8019]
PCA 변환 신용카드 고객 dataset 평균 정확도:  0.796
PCA Component 별 변동성:  [0.28448215 0.17818817 0.06743307 0.06401154 0.04457555 0.04161736]


두 결과를 보자. PCA 변환을 적용한 전체 피처의 1/4 정도만으로도 예측 성능이 PCA 변환을 적용하지 않은 원본 데이터와 크게 차이가 나지 않는다.<br>
이러면에서 PCA는 유용하다 볼 수 있지만, 원본 데이터의 변동성이 65 % 수준에 머무르는 것을 보면 항상 유용하다 보기는 어려운 것 같다. <br><br>
이 부분은 관련 논문을 찾아본 뒤 리뷰를 하면서 자세하게 알아 보도록 하자. <br><br>
추가로 PCA 는 computer vision 분야에서 더 활발하게 사용된다. 